crypto/async/async_wait.c

```diff
@@ -148,6 +148,33 @@ int ASYNC_WAIT_CTX_clear_fd(ASYNC_WAIT_CTX *ctx, const void *key)
             continue;
         }
         if (curr->key == key) {
+            /* If fd has just been added, remove it from the list */
+            if (curr->add == 1) {
+                if (ctx->fds == curr) {
+                    ctx->fds = curr->next;
+                }
+                else {
+                    struct fd_lookup_st *prev = ctx->fds;
+                    while (prev->next != curr && prev->next != NULL) {
+                        prev = prev->next;
+                    }
+                    if (prev->next == NULL) {
+                        return 1;
+                    }
+                    prev->next = curr->next;
+                }
+
+                /*
+                 * The fd has just been added so it can't be used externally
+                 * and it is safe to call the cleanup function here
+                 */
+                if (curr->cleanup != NULL)
+                    curr->cleanup(ctx, curr->key, curr->fd, curr->custom_data);
```

Comment:
Hmmm. I'm not sure this is a good idea. It kind of confuses the semantics of "cleanup". At the moment the documentation says:
"The "cleanup" routine does not get called if the fd is cleared directly via a call to ASYNC_WAIT_CTX_clear_fd()"